packages/fxa-settings/src/components/App/index.tsx

```diff
@@ -32,7 +37,42 @@ import SigninReported from '../../pages/Signin/SigninReported';
 export const App = ({
   flowQueryParams,
 }: { flowQueryParams: QueryParams } & RouteComponentProps) => {
+  const [isSignedIn, setIsSignedIn] = useState<boolean>();
+
   const { showReactApp } = flowQueryParams;
+  const { loading, error } = useInitialState();
+  const { metricsEnabled } = useAccount();
+  const config = useConfig();
+
+  useEffect(() => {
+    Metrics.init(metricsEnabled, flowQueryParams);
+  }, [metricsEnabled, flowQueryParams]);
+
+  useEffect(() => {
+    if (!loading && error?.message.includes('Invalid token')) {
+      setIsSignedIn(false);
+    } else if (!loading && !error) {
+      setIsSignedIn(true);
+    }
+  }, [error, loading]);
+
+  useEffect(() => {
+    if (!loading) {
+      // For settings app, we only enable Sentry once we know the user's metrics
+      // preferences. A bit of chicken and egg but it could be possible that we
+      // miss some errors while the page is loading and user is being fetched.
+      if (metricsEnabled || !isSignedIn) {
```

Comment:
I kind of wonder if we should have another check in Settings when we know the user is logged in that checks if Sentry has been initialized and if it hasn't, check again against `metricsEnabled` and initialize if they're opted in. Otherwise, I think if a user has opted out, we'll initialize when they aren't signed in, and when they login, we'll collect metrics regardless of status? Maybe the ideal scenario is to perform this check on successful login instead of within Settings since there's other pages besides Settings that users can be logged in for, but doing this check in Settings in the interim seems fine.

@chenba may be able to do this in [FXA-7006](https://mozilla-hub.atlassian.net/browse/FXA-7006) since whatever we do we probably want to do the same checks in that other `useEffect` with `Metrics.init`.

Reply:
I'm still having a hard time understanding what changes I need to make.  The code seems to be fine.  When the `isSignedIn` state changes, the component re-renders, executing the hooks, and things enable/disable accordingly.  What am I missing?

Reply:
@chenba ah, maybe we're good then - I thought the way we were initializing our metrics in `metrics.ts` would force us to perform some extra logic after the user had logged in/out. If this hook always reruns when the user signed in status changes from whatever component, then that seems fine. Thanks for double checking me. 😄